episodes/testing.md

```diff
@@ -0,0 +1,245 @@
+ - Use runtime testing to prevent weird cases. (hint to simulation testing?)
+::::::
+
+# Introduction
+
+In this episode we are going to take a look at a few different types of automated testing. We will also see how we can use code coverage the increase our confidence that everything still works when we make a change to the code. There is an assumed base of having worked through the material on [this website](https://coderefinery.github.io/testing/motivation/). We will also make use of the [python template](https://github.com/SS-NES/python-template) to get a standardized starting point.
+
+
+# 1. Improve testing
+
+## Add code coverage
+
+What is code coverage? Code coverage is the percentage of the research / production code you have written that is covered by unittests. If you have a very high percentage then when you make a change in the code, and it breaks the chances of it being caught before showing it to users is very high. If you have a low percentage the chances of finding these bugs are low, or you have to do a lot of manual testing.  When working with python and pytest there are packages to easily get the test coverage of your application. The one used most is [pytest-cov](https://pytest-cov.readthedocs.io/en/latest/config.html) .
+
+## Add parameterized tests
+
+When writing tests it sometimes happens that you want a lot of tests for the same function. You could write a lot of test functions with the same setup and when calling the function under test some different parameters. A cleaner way where you have to maintain less code afterwards to do this is by using paramterized tests. With this you add the different parameters as inputs to you test function. An example of this looks like this:
+
+```python
+@pytest.mark.parametrize(
+    ("onset", "phenomenon", "expected"),
+    [
+        (
+            "2024-12-09T11:31:14Z",
+            "snow-ice",
+            "Monday 9 December: chance of snow/road icing",
+        ),
+        (
+            "2025-01-04T00:00:00Z",
+            "low-temperature",
+            "Saturday 4 January: chance of cold",
+        ),
+    ],
+    ids=["special_case", "normal_case"],
+)
+def test_get_english_headline(onset: str, phenomenon: str, expected: str) -> None:
+    """test generation of english headline"""
+    assert _get_english_headline({"onset": onset, "phenomenon": phenomenon}) == expected
+```
+
+As you can see even the expected result is now an input of the test. We can use the ids parameter to give a test a name. With this name you can also run the test for only one of the ids.
+
+
+# 2. Testing a unit of software without having to instantiate the all the code
+
+Sometimes it happens that you want to test function but in it a lot of complex objects are used that also need other objects. One way to deal with this is to add those complex objects as input to the function. You can that use this mock to prevent you having to create all those objects yourself.
```

Comment:
```suggestion
Sometimes it happens that you want to test a function but in that function a lot of complex objects are used (and those objects in turn need other objects...). One way to deal with this is to add those complex objects as input to the function. You can that use this mock to prevent you having to create all those objects yourself.
```